cl/lib/elasticsearch_utils.py

```diff
@@ -2679,6 +2685,73 @@ def apply_custom_score_to_main_query(
     return query
 
 
+def build_semantic_query(
+    text_query: str, fields: list[str], filters: list[QueryString | Range]
+) -> tuple[str, list[Query]]:
+    """
+    Build a hybrid Elasticsearch query using both exact keyword matching and
+    semantic vector search.
+
+    :param text_query: The raw user query string, which may include quoted
+        phrases for exact matching.
+    :param fields: A list of fields to target with the full-text keyword query.
+    :param filters: A list of filter clauses to apply as pre-filtering to the
+        semantic KNN search query.
+    :return: A two-tuple:
+        - keyword_query: A string representing the AND-joined quoted phrases, if any.
+        - semantic_query: A list of Elasticsearch Q objects, including the KNN vector search
+          and optionally a keyword-based full-text query.
+    :raises InputTooLongError: If the cleaned query string exceeds the maximum allowed length
+        for generating embeddings.
+    """
+    semantic_query: list[Query] = []
+    # Extract quoted phrases from the input string (for exact keyword matching)
+    exact_keywords = re.findall(r'"([^"]*)"', text_query)
+
+    # Join extracted phrases with AND to form a keyword query string
+    keyword_query = " AND ".join([f'"{s}"' for s in exact_keywords])
```

Comment:
It's not necessary to join the exact terms with `AND`, since the resulting query string uses `AND` as the default operator.


```
{
    "query_string":{
       "query":"\"cases before\"  \"questions which\"",
       "default_operator":"AND",
       "fields":[
          "court",
          "court_id",
          "citation",
          "judge",
          "caseNameFull",
          "status",
          "suitNature",
          "attorney",
          "procedural_history",
          "posture",
          "syllabus",
          "text^1.0",
          "type^1.0",
          "caseName.exact^4.0",
          "docketNumber^2.0"
       ],
       "fuzziness":2,
       "quote_field_suffix":".exact",
       "tie_breaker":0.3
    }
 }
```